### DISCUSSION TITLE
Getting Run Path Message When Training

### DISCUSSION THREAD

**Discussion by carson2stoker** [talmolab/sleap#832] [Help!]
I am attempting to train a single instance model using the GUI after some initial labeling. Every time I hit hit run, training begins but never makes it past the first epoch. I always get a dialog box saying 'An error occurred while training single_instance. Your command line terminal may have more information about the error' and this message in the terminal. 
```
Epoch 1/200
Run Path: /uufs/chpc.utah.edu/common/home/u1208563/Desktop/SLEAP/models/220708_133349.single_instance.n=8
```

Once I click OK on the dialog box, this also shows. 
```
qt.qpa.xcb: QXcbConnection: XCB error: 3 (BadWindow), sequence: 7734, resource id: 6291472, major code: 40 (TranslateCoords), minor code: 0
```

<details><summary>Logs</summary>

```
(base) [u1208563@notchpeak1:~]$ salloc --time=3:00:00 --account=notchpeak-gpu --partition=notchpeak-gpu --nodes=1 --ntasks=1 --mem=3000  --gres=gpu:1
salloc: Pending job allocation 4907373
salloc: job 4907373 queued and waiting for resources
salloc: job 4907373 has been allocated resources
salloc: Granted job allocation 4907373
salloc: Waiting for resource configuration
salloc: Nodes notch328 are ready for job
(base) [u1208563@notch328:~]$ module use $HOME/MyModules
(base) [u1208563@notch328:~]$ module load miniconda3/latest
(base) [u1208563@notch328:~]$ conda activate sleap
(sleap) [u1208563@notch328:~]$ sleap-label
Saving config: /uufs/chpc.utah.edu/common/home/u1208563/.sleap/1.2.4/preferences.yaml
QStandardPaths: error creating runtime directory /run/user/1292742 (Permission denied)
Restoring GUI state...

Software versions:
SLEAP: 1.2.4
TensorFlow: 2.6.3
Numpy: 1.19.5
Python: 3.7.12
OS: Linux-4.18.0-348.20.1.el8_5.x86_64-x86_64-with-centos-8.5-Green_Obsidian

Happy SLEAPing! :)
[aac @ 0x55d2262d3d00] Input buffer exhausted before END element found
Resetting monitor window.
Polling: /uufs/chpc.utah.edu/common/home/u1208563/Desktop/SLEAP/models/220708_133349.single_instance.n=8/viz/validation.*.png
Start training single_instance...
['sleap-train', '/scratch/local/u1208563/4907373/tmprfjmkzhl/220708_133349_training_job.json', '/uufs/chpc.utah.edu/common/home/u1208563/Desktop/SLEAP/labels.v000.slp', '--zmq', '--save_viz']
qt.qpa.xcb: QXcbConnection: XCB error: 3 (BadWindow), sequence: 4025, resource id: 6291468, major code: 40 (TranslateCoords), minor code: 0
INFO:sleap.nn.training:Versions:
SLEAP: 1.2.4
TensorFlow: 2.6.3
Numpy: 1.19.5
Python: 3.7.12
OS: Linux-4.18.0-348.20.1.el8_5.x86_64-x86_64-with-centos-8.5-Green_Obsidian
INFO:sleap.nn.training:Training labels file: /uufs/chpc.utah.edu/common/home/u1208563/Desktop/SLEAP/labels.v000.slp
INFO:sleap.nn.training:Training profile: /scratch/local/u1208563/4907373/tmprfjmkzhl/220708_133349_training_job.json
INFO:sleap.nn.training:
INFO:sleap.nn.training:Arguments:
INFO:sleap.nn.training:{
    "training_job_path": "/scratch/local/u1208563/4907373/tmprfjmkzhl/220708_133349_training_job.json",
    "labels_path": "/uufs/chpc.utah.edu/common/home/u1208563/Desktop/SLEAP/labels.v000.slp",
    "video_paths": [
        ""
    ],
    "val_labels": null,
    "test_labels": null,
    "tensorboard": false,
    "save_viz": true,
    "zmq": true,
    "run_name": "",
    "prefix": "",
    "suffix": "",
    "cpu": false,
    "first_gpu": false,
    "last_gpu": false,
    "gpu": 0
}
INFO:sleap.nn.training:
INFO:sleap.nn.training:Training job:
INFO:sleap.nn.training:{
    "data": {
        "labels": {
            "training_labels": null,
            "validation_labels": null,
            "validation_fraction": 0.1,
            "test_labels": null,
            "split_by_inds": false,
            "training_inds": null,
            "validation_inds": null,
            "test_inds": null,
            "search_path_hints": [],
            "skeletons": []
        },
        "preprocessing": {
            "ensure_rgb": false,
            "ensure_grayscale": false,
            "imagenet_mode": null,
            "input_scaling": 1.0,
            "pad_to_stride": null,
            "resize_and_pad_to_target": true,
            "target_height": null,
            "target_width": null
        },
        "instance_cropping": {
            "center_on_part": null,
            "crop_size": null,
            "crop_size_detection_padding": 16
        }
    },
    "model": {
        "backbone": {
            "leap": null,
            "unet": {
                "stem_stride": null,
                "max_stride": 16,
                "output_stride": 2,
                "filters": 16,
                "filters_rate": 2.0,
                "middle_block": true,
                "up_interpolate": true,
                "stacks": 1
            },
            "hourglass": null,
            "resnet": null,
            "pretrained_encoder": null
        },
        "heads": {
            "single_instance": {
                "part_names": null,
                "sigma": 2.5,
                "output_stride": 2,
                "loss_weight": 1.0,
                "offset_refinement": false
            },
            "centroid": null,
            "centered_instance": null,
            "multi_instance": null,
            "multi_class_bottomup": null,
            "multi_class_topdown": null
        }
    },
    "optimization": {
        "preload_data": true,
        "augmentation_config": {
            "rotate": true,
            "rotation_min_angle": -15.0,
            "rotation_max_angle": 15.0,
            "translate": false,
            "translate_min": -5,
            "translate_max": 5,
            "scale": false,
            "scale_min": 0.9,
            "scale_max": 1.1,
            "uniform_noise": false,
            "uniform_noise_min_val": 0.0,
            "uniform_noise_max_val": 10.0,
            "gaussian_noise": false,
            "gaussian_noise_mean": 5.0,
            "gaussian_noise_stddev": 1.0,
            "contrast": false,
            "contrast_min_gamma": 0.5,
            "contrast_max_gamma": 2.0,
            "brightness": false,
            "brightness_min_val": 0.0,
            "brightness_max_val": 10.0,
            "random_crop": false,
            "random_crop_height": 256,
            "random_crop_width": 256,
            "random_flip": false,
            "flip_horizontal": true
        },
        "online_shuffling": true,
        "shuffle_buffer_size": 128,
        "prefetch": true,
        "batch_size": 4,
        "batches_per_epoch": null,
        "min_batches_per_epoch": 200,
        "val_batches_per_epoch": null,
        "min_val_batches_per_epoch": 10,
        "epochs": 200,
        "optimizer": "adam",
        "initial_learning_rate": 0.0001,
        "learning_rate_schedule": {
            "reduce_on_plateau": true,
            "reduction_factor": 0.5,
            "plateau_min_delta": 1e-06,
            "plateau_patience": 5,
            "plateau_cooldown": 3,
            "min_learning_rate": 1e-08
        },
        "hard_keypoint_mining": {
            "online_mining": false,
            "hard_to_easy_ratio": 2.0,
            "min_hard_keypoints": 2,
            "max_hard_keypoints": null,
            "loss_scale": 5.0
        },
        "early_stopping": {
            "stop_training_on_plateau": true,
            "plateau_min_delta": 1e-08,
            "plateau_patience": 10
        }
    },
    "outputs": {
        "save_outputs": true,
        "run_name": "220708_133349.single_instance.n=8",
        "run_name_prefix": "",
        "run_name_suffix": "",
        "runs_folder": "/uufs/chpc.utah.edu/common/home/u1208563/Desktop/SLEAP/models",
        "tags": [
            ""
        ],
        "save_visualizations": true,
        "delete_viz_images": true,
        "zip_outputs": false,
        "log_to_csv": true,
        "checkpointing": {
            "initial_model": false,
            "best_model": true,
            "every_epoch": false,
            "latest_model": false,
            "final_model": false
        },
        "tensorboard": {
            "write_logs": false,
            "loss_frequency": "epoch",
            "architecture_graph": false,
            "profile_graph": false,
            "visualizations": true
        },
        "zmq": {
            "subscribe_to_controller": true,
            "controller_address": "tcp://127.0.0.1:9000",
            "controller_polling_timeout": 10,
            "publish_updates": true,
            "publish_address": "tcp://127.0.0.1:9001"
        }
    },
    "name": "",
    "description": "",
    "sleap_version": "1.2.4",
    "filename": "/scratch/local/u1208563/4907373/tmprfjmkzhl/220708_133349_training_job.json"
}
INFO:sleap.nn.training:
INFO:sleap.nn.training:Using GPU 0 for acceleration.
INFO:sleap.nn.training:Disabled GPU memory pre-allocation.
INFO:sleap.nn.training:System:
GPUs: 1/1 available
  Device: /physical_device:GPU:0
         Available: True
        Initalized: False
     Memory growth: True
INFO:sleap.nn.training:
INFO:sleap.nn.training:Initializing trainer...
INFO:sleap.nn.training:Loading training labels from: /uufs/chpc.utah.edu/common/home/u1208563/Desktop/SLEAP/labels.v000.slp
INFO:sleap.nn.training:Creating training and validation splits from validation fraction: 0.1
INFO:sleap.nn.training:  Splits: Training = 7 / Validation = 1.
INFO:sleap.nn.training:Setting up for training...
INFO:sleap.nn.training:Setting up pipeline builders...
INFO:sleap.nn.training:Setting up model...
INFO:sleap.nn.training:Building test pipeline...
[aac @ 0x557400a74800] Input buffer exhausted before END element found
2022-07-08 13:33:54.822144: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-07-08 13:33:57.272478: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1510] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 22319 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 3090, pci bus id: 0000:c1:00.0, compute capability: 8.6
2022-07-08 13:33:57.852577: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] None of the MLIR Optimization Passes are enabled (registered 2)
INFO:sleap.nn.training:Loaded test example. [5.309s]
INFO:sleap.nn.training:  Input shape: (1088, 1920, 1)
INFO:sleap.nn.training:Created Keras model.
INFO:sleap.nn.training:  Backbone: UNet(stacks=1, filters=16, filters_rate=2.0, kernel_size=3, stem_kernel_size=7, convs_per_block=2, stem_blocks=0, down_blocks=4, middle_block=True, up_blocks=3, up_interpolate=True, block_contraction=False)
INFO:sleap.nn.training:  Max stride: 16
INFO:sleap.nn.training:  Parameters: 1,953,765
INFO:sleap.nn.training:  Heads: 
INFO:sleap.nn.training:    [0] = SingleInstanceConfmapsHead(part_names=['head', 'thorax', 'Lant1', 'Lant2', 'Rant1', 'Rant2', 'fLleg1', 'fLleg2', 'fRleg1', 'fRleg2', 'mLleg1', 'mLleg2', 'mRleg1', 'mRleg2', 'hLleg1', 'hLleg2', 'hRleg1', 'hRleg2', 'abdo', 'Rwing', 'Lwing'], sigma=2.5, output_stride=2, loss_weight=1.0)
INFO:sleap.nn.training:  Outputs: 
INFO:sleap.nn.training:    [0] = KerasTensor(type_spec=TensorSpec(shape=(None, 544, 960, 21), dtype=tf.float32, name=None), name='SingleInstanceConfmapsHead/BiasAdd:0', description="created by layer 'SingleInstanceConfmapsHead'")
INFO:sleap.nn.training:Setting up data pipelines...
INFO:sleap.nn.training:Training set: n = 7
INFO:sleap.nn.training:Validation set: n = 1
INFO:sleap.nn.training:Setting up optimization...
INFO:sleap.nn.training:  Learning rate schedule: LearningRateScheduleConfig(reduce_on_plateau=True, reduction_factor=0.5, plateau_min_delta=1e-06, plateau_patience=5, plateau_cooldown=3, min_learning_rate=1e-08)
INFO:sleap.nn.training:  Early stopping: EarlyStoppingConfig(stop_training_on_plateau=True, plateau_min_delta=1e-08, plateau_patience=10)
INFO:sleap.nn.training:Setting up outputs...
INFO:sleap.nn.callbacks:Training controller subscribed to: tcp://127.0.0.1:9000 (topic: )
INFO:sleap.nn.training:  ZMQ controller subcribed to: tcp://127.0.0.1:9000
INFO:sleap.nn.callbacks:Progress reporter publishing on: tcp://127.0.0.1:9001 for: not_set
INFO:sleap.nn.training:  ZMQ progress reporter publish on: tcp://127.0.0.1:9001
INFO:sleap.nn.training:Created run path: /uufs/chpc.utah.edu/common/home/u1208563/Desktop/SLEAP/models/220708_133349.single_instance.n=8
INFO:sleap.nn.training:Setting up visualization...
INFO:sleap.nn.training:Finished trainer set up. [6.4s]
INFO:sleap.nn.training:Creating tf.data.Datasets for training data generation...
INFO:sleap.nn.training:Finished creating training datasets. [5.0s]
INFO:sleap.nn.training:Starting training loop...
Epoch 1/200
Run Path: /uufs/chpc.utah.edu/common/home/u1208563/Desktop/SLEAP/models/220708_133349.single_instance.n=8
qt.qpa.xcb: QXcbConnection: XCB error: 3 (BadWindow), sequence: 7734, resource id: 6291472, major code: 40 (TranslateCoords), minor code: 0

```

</details>

I'm still trying to learn SLEAP, so it's very possible I just made a dumb mistake. I have some programming experience, but not really with this kind of thing before. So I really appreciate any help an am happy to provide more info if needed. Thanks!

Carson

**Comment by talmo** [comment]
Hi @carson2stoker,

This looks like it might be an issue with the cluster configuration. It looks like you're connecting to a remote server in which you're running the GUI -- this might not work since graphical applications tend not to play nicely with remote graphics emulation.

Do you mind trying to [train on Colab](https://sleap.ai/notebooks/Training_and_inference_using_Google_Drive.html) as a check to see if it's a SLEAP issue or an environment issue?

If it works on there, the workaround is to label on your own computer, export the training package and train using the CLI in your cluster, but let's eliminate the possibility that it's a SLEAP/configuration issue first.

Cheers,

Talmo

**Reply by carson2stoker** [reply]
I thought that was the point of using an interactive job, that it was supposed to allow you to work with GUIs. Is that incorrect? Colab was actually the original route I tried to take before using the cluster. When using Colab, I did not experience that same RUN PATH message that I did on the cluster, if that answers your question. However, it was really slow and I had trouble with exporting the predictions back into the GUI. To be honest, the documentation is unclear on how to continue using colab after the initial training and inference are done. But I was able to complete the initial training and inference on colab without any trouble. So maybe labeling on my computer and then training using the CLI in the cluster is the way to go.
